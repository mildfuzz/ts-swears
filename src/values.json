{
  "ddl": {
    "page": {
      "nbs_environment": null,
      "nbs_server": null,
      "pageInfo": {
        "pageName": null,
        "nbs_page_load_time": null,
        "nbs_page_responsive_state": null,
        "nbs_page_responsive_orientation": null
      },
      "category": {
        "primaryCategory": "over payment application",
        "nbs_sub_category_1": null,
        "nbs_sub_category_2": null,
        "nbs_sub_category_3": null
      }
    },

    "nbs_message": [],
    "nbs_user_input": [],
    "nbs_element_interaction": [],
    "nbs_app_mortgage": {},

    "nbs_user": {
      "nbs_user_application_route": null
    }
  },

  "regionProperties": {
    "view": ["page"],
    "action": ["nbs_user_input", "nbs_element_interaction", "nbs_message"],
    "context": ["nbs_user", "nbs_app_mortgage"]
  },

  "itemsKeyMap": {
    "nbs_message": {
      "category": null,
      "text": null,
      "cause": null
    },

    "nbs_user_input": {
      "name": "nbs_input_name",
      "status": "nbs_input_validation_status",
      "message": "nbs_input_validation_message"
    },

    "nbs_element_interaction": {
      "type": "nbs_interaction_type",
      "label": "nbs_interaction_label"
    }
  },

  "page": {
    "PREFIX": "nbs-aws",
    "TYPE": "application",
    "APP_NAME": "OverPayment",
    "INITIAL": "initial load",

    "types": {
      "LOADER": "loader"
    },
    "orientations": {
      "PORTRAIT": "portrait",
      "LANDSCAPE": "landscape",
      "SQUARE": "square"
    },
    "sizes": {
      "SMALL": "small",
      "MEDIUM": "medium",
      "LARGE": "large"
    }
  },

  "valueMaps": {
    "environments": {
      "home-mumo-dev": "dev",
      "home-mumo-prd": "live",
      "home-afford-dev-rtl": "dev",
      "home-afford-sit-rtl": "sit"
    },

    "routes": {
      "referrers": {
        "://(www.)?nationwide.co.uk": "nationwide.co.uk",
        "://onlinebanking.nationwide.co.uk": "internet bank",
        "^$": "external"
      },
      "urlParams": {
        "JId": "banking app"
      }
    },

    "inputValidity": {
      "true": "passed",
      "false": "failed",
      "null": "none"
    }
  },

  "types": {
    "action": {
      "SUBMIT": "submit",
      "INPUT": "user_input_complete",
      "DEFAULT": "misc_interaction"
    },

    "interaction": {
      "?LINK": "external link",
      "SUBMIT": "submit",
      "REVEAL": "reveal",
      "HELP": "help icon",
      "PRINT": "print",
      "PDF": "save pdf",
      "CALL": "click to call",
      "BACK": "back",
      "MISC": "misc"
    },

    "message": {
      "USER": "user error",
      "BUSINESS": "business error",
      "SYSTEM": "system error",
      "NOTIFICATION": "notification"
    }
  }
}
