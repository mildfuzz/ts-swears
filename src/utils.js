import _get from 'lodash/get'
import _map from 'lodash/map'
import _join from 'lodash/join'
import _trim from 'lodash/trim'
import _take from 'lodash/take'
import _find from 'lodash/find'
import _round from 'lodash/round'
import _concat from 'lodash/concat'
import _reduce from 'lodash/reduce'
import _compact from 'lodash/compact'
import _findKey from 'lodash/findKey'
import _toLower from 'lodash/toLower'
import _includes from 'lodash/includes'
import _isObject from 'lodash/isObject'
import _isUndefined from 'lodash/isUndefined'

import values from './values.json'

/**
 * Page name formatting
 * Tagging: Spec - 4.5.1.3
 */
export const getDdlText = (text) => _trim(_toLower(text || ''))

/**
 * Returns page name string from name sections array
 * Tagging: Spec - 4.5.1.3
 *
 * @param {array} sections
 * @returns {string} lowercase concatenated name sections
 */
export const toDdlName = (sections) =>
  _join(_compact(_map(sections, getDdlText)), ':')

/**
 * Returns page name sections using global and page specific sections
 * Tagging: Spec - 4.5.1.3
 *
 * @param {string|array} desc
 * @returns {array} page name sections
 */
export const getPageNameSections = (desc) => {
  const {PREFIX, TYPE, APP_NAME, APP_SUBNAME} = values.page
  return _concat(_compact([PREFIX, TYPE, APP_NAME, APP_SUBNAME]), desc)
}

export const getPageName = (desc) => toDdlName(getPageNameSections(desc))

export const getPageCategories = (desc) => {
  const pageNameSections = getPageNameSections(desc)

  return {
    nbs_sub_category_1: toDdlName(_take(pageNameSections, 2)),
    nbs_sub_category_2: toDdlName(_take(pageNameSections, 3)),
    nbs_sub_category_3: toDdlName(_take(pageNameSections, 4)),
  }
}

/**
 * Returns the region for ddl property
 * @param {string} property
 * @returns {string} region
 */
export const getPropertyRegion = (property) =>
  _findKey(
    values.regionProperties,
    (properties, region) => _includes(properties, property) && region,
  )

/**
 * Checks if property is context specific
 * @param {string} property
 * @return {bool}
 */
export const isContextProperty = (property) =>
  getPropertyRegion(property) === 'context'

/**
 * Checks if property is action specific
 * @param {string} property
 * @return {bool}
 */
export const isActionProperty = (property) =>
  getPropertyRegion(property) === 'action'

/**
 * Returns the key for a DDL item, if item or key not found returns key provided
 * @param {string} property
 * @param {string} item
 * @returns {string} DDL item property key
 */
export const getItemPropertyKey = (item, property) =>
  _get(values.itemsKeyMap, [item, property], property) || property

/**
 * Load time measuring
 * Tagging: Spec - 5.1.6.1, Req. - 1.2
 */
export const pageLoadTimes = {
  [getPageName(values.page.INITIAL)]: global.initialLoadStart,
}

const timeNow = () =>
  global?.performance?.now ? global?.performance?.now() : Date?.now()

export const loadStart = (pageName) => {
  const loadTime = timeNow()
  pageLoadTimes[pageName] = loadTime
  return loadTime
}

export const loadEnd = (pageName) => {
  const loadTime = _round(Math.abs(timeNow() - pageLoadTimes[pageName]))
  pageLoadTimes[pageName] = loadTime
  return loadTime
}

export const getEntryRoute = () => {
  const {
    referrer,
    location: {search},
  } = global.document
  const {referrers, urlParams} = values.valueMaps.routes
  const defaultRoute = referrers['^$']
  let entryRoute = _find(referrers, (_, pattern) =>
    new RegExp(pattern).test(referrer),
  )

  if (!entryRoute || entryRoute === defaultRoute) {
    entryRoute =
      _find(urlParams, (_, pattern) =>
        new RegExp(`[\\?&]${pattern}=`, 'i').test(search),
      ) || entryRoute
  }

  return entryRoute || defaultRoute
}

/**
 * Technical service data
 * Tagging: Spec - 5.1.6.1, Req. - 1.2
 */
export const getServiceData = (env) => ({
  nbs_environment: env,
  nbs_server: global.location.origin,
})

/**
 * Device data
 * Tagging: Spec - 5.1.6.1, Req. - 1.3
 */
export const getDeviceData = ({width, height}) => {
  const {orientations, sizes} = values.page

  let orientation
  let state

  switch (width % height) {
    case 0:
      orientation = orientations.SQUARE
      break
    case width:
      orientation = orientations.PORTRAIT
      break
    default:
      orientation = orientations.LANDSCAPE
      break
  }

  if (width <= 480) state = sizes.SMALL
  else if (width <= 768) state = sizes.MEDIUM
  else state = sizes.LARGE

  return {
    nbs_page_responsive_orientation: orientation,
    nbs_page_responsive_state: state,
  }
}

/**
 * converts formik error data structure to analytics message structure
 */
export const formatFormikErrors = (formikErrors) => {
  const getMappedErrors = (err, curPath) =>
    _reduce(
      err,
      (acc, val, key) => {
        const path = !_isUndefined(curPath) ? `${curPath}.${key}` : key

        if (_isObject(val)) {
          return {...acc, ...getMappedErrors(val, path)}
        }

        return {
          ...acc,
          [path]: val,
        }
      },
      {},
    )

  const errors = getMappedErrors(formikErrors)

  return _map(errors, (message, cause) => ({message, cause}))
}
