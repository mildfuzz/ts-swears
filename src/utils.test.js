import _map from 'lodash/map'
import _clone from 'lodash/clone'
import _forEach from 'lodash/forEach'

import values from './values.json'
import * as utils from './utils'

describe('@analytics > utils', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('page name formatting', () => {
    describe('getDdlText()', () => {
      test('returns trimmed and lowercase text', () => {
        _forEach(
          [
            {text: ' ', expected: ''},
            {text: 0, expected: ''},
            {text: null, expected: ''},
            {text: false, expected: ''},
            {text: undefined, expected: ''},
            {text: 1, expected: '1'},
            {text: ' Foo ', expected: 'foo'},
            {text: 'FOO ', expected: 'foo'},
            {text: ' fOo', expected: 'foo'},
            {text: ' foo Bar ', expected: 'foo bar'},
          ],
          ({text, expected}) => {
            expect(utils.getDdlText(text)).toEqual(expected)
          },
        )
      })
    })

    describe('toDdlName()', () => {
      test('returns lowercase & trimmed delimited string from array', () => {
        _forEach(
          [
            {text: ' ', expected: ''},
            {
              text: ['', ' ', 0, 1, '0', null, false, undefined],
              expected: '1:0',
            },
            {text: [' FOO '], expected: 'foo'},
            {text: [' Foo ', 'baR'], expected: 'foo:bar'},
            {text: [' FOO ', '   ', ' BAR '], expected: 'foo:bar'},
          ],
          ({text, expected}) => {
            expect(utils.toDdlName(text)).toEqual(expected)
          },
        )
      })
    })

    describe('getPageNameSections()', () => {
      test('returns page name sections array with category prefixes', () => {
        expect(
          _map(
            [0, null, false, undefined, '', [0, 1, '0', 'foo', 'bar']],
            utils.getPageNameSections,
          ),
        ).toMatchSnapshot()
      })
    })

    describe('getPageName()', () => {
      test('returns lowercase & trimmed page name string', () => {
        expect(
          _map(
            [0, null, false, undefined, '', [0, 1, '0', 'foo', 'bar']],
            utils.getPageName,
          ),
        ).toMatchSnapshot()
      })
    })
  })

  describe('page categories', () => {
    describe('getPageCategories()', () => {
      test('returns correctly constructed object', () => {
        expect(
          _map(['', 'foo 1', ['bar 1', 'bar 2']], utils.getPageCategories),
        ).toMatchSnapshot()
      })
    })
  })

  describe('ddl property', () => {
    describe('getPropertyRegion()', () => {
      test('return property region name for known property', () => {
        _forEach(values.regionProperties, (properties, region) => {
          _forEach(properties, (property) => {
            expect(utils.getPropertyRegion(property)).toEqual(region)
          })
        })
      })

      test('return undefined for unknown property', () => {
        expect(utils.getPropertyRegion('unknown')).toEqual()
      })
    })

    describe('isContextProperty()', () => {
      test('return true only if provided property belongs to the context region', () => {
        _forEach(values.regionProperties.context, (property) => {
          expect(utils.isContextProperty(property)).toBe(true)
        })

        _forEach(values.regionProperties.view, (property) => {
          expect(utils.isContextProperty(property)).toBe(false)
        })

        expect(utils.isContextProperty('unknown')).toBe(false)
      })
    })

    describe('isActionProperty()', () => {
      test('return true only if provided property belongs to the action region', () => {
        _forEach(values.regionProperties.action, (property) => {
          expect(utils.isActionProperty(property)).toBe(true)
        })

        _forEach(values.regionProperties.context, (property) => {
          expect(utils.isActionProperty(property)).toBe(false)
        })

        expect(utils.isActionProperty('unknown')).toBe(false)
      })
    })

    describe('getItemPropertyKey()', () => {
      test('return the ddl property item key from items key forEach if available', () => {
        _forEach(values.itemsKeyMap, (items, item) => {
          _forEach(items, (propertyKey, property) => {
            expect(utils.getItemPropertyKey(item, property)).toBe(
              propertyKey || property,
            )
          })
        })
      })
    })
  })

  describe('load time measuring', () => {
    const mockDateTime = 111
    const mockPerfTime = 222

    const getUtils = () => require('./utils') // eslint-disable-line global-require

    beforeEach(() => {
      jest.resetModules()
    })

    jest.spyOn(global.Date, 'now').mockReturnValue(mockDateTime)

    test('initial load start time', () => {
      global.initialLoadStart = 101
      expect(getUtils().pageLoadTimes).toMatchSnapshot()
    })

    describe('loadStart()', () => {
      test('sets & returns page load start time', () => {
        global.performance.now = jest.fn().mockReturnValue(mockPerfTime)
        const {loadStart, pageLoadTimes} = getUtils()
        expect(loadStart('foo')).toEqual(mockPerfTime)

        global.performance.now = null
        expect(loadStart('bar')).toEqual(mockDateTime)

        expect(pageLoadTimes).toMatchSnapshot()
      })
    })

    describe('loadEnd()', () => {
      test('sets & returns page load duration', () => {
        global.performance.now = jest.fn().mockReturnValue(mockPerfTime)
        const {loadStart, loadEnd, pageLoadTimes} = getUtils()
        const loadTimes = {}

        loadStart('foo')
        loadTimes.withPerf = {start: _clone(pageLoadTimes)}

        global.performance.now = jest.fn().mockReturnValue(mockPerfTime + 10)
        expect(loadEnd('foo')).toEqual(10)
        loadTimes.withPerf.end = _clone(pageLoadTimes)

        global.performance.now = null
        loadStart('bar')
        loadTimes.withDate = {start: _clone(pageLoadTimes)}

        global.Date.now.mockReturnValue(mockDateTime + 10)
        expect(loadEnd('bar')).toEqual(10)
        loadTimes.withDate.end = _clone(pageLoadTimes)

        expect(loadTimes).toMatchSnapshot()
      })
    })
  })

  describe('getServiceData()', () => {
    test('returns service properties with mapped environment name', () => {
      expect(
        _map(['local', 'dev', 'live', 'unknown'], utils.getServiceData),
      ).toMatchSnapshot()
    })
  })

  describe('getDeviceData()', () => {
    test('returns device data with orientation & responsive state', () => {
      expect(
        _map(
          [
            {width: 1, height: 1}, // square ratio
            {width: 300, height: 300}, // square small resolution

            {width: 3, height: 4}, // portrait (tablet) ratio
            {width: 768, height: 1024}, // portrait (tablet) medium resolution

            {width: 9, height: 16}, // portrait (mobile) ratio
            {width: 720, height: 1280}, // portrait (mobile) medium resolution

            {width: 5, height: 4}, // landscape ratio
            {width: 1280, height: 1024}, // landscape large resolution

            {width: 8, height: 5}, // landscape (monitor)
            {width: 1680, height: 1050}, // landscape (monitor) large resolution

            {width: 16, height: 9}, // landscape (widescreen) ratio
            {width: 1920, height: 1080}, // landscape (widescreen) large resolution
          ],
          ({width, height}) => [
            `${width}x${height}`,
            utils.getDeviceData({width, height}),
          ],
        ),
      ).toMatchSnapshot()
    })
  })

  describe('getEntryRoute()', () => {
    const referrers = [
      'http://nationwide.co.uk/abc-123',
      'http://www.nationwide.co.uk/abc-123?xyz=123',
      'https://nationwide.co.uk?abc-123',
      'https://www.nationwide.co.uk/abc-123',
      'http://onlinebanking.nationwide.co.uk?GSSToken=abc-123',
      'https://onlinebanking.nationwide.co.uk/abc-123/?GSSToken=abc-123',
      'https://google.co.uk/search',
      '',
      undefined,
    ]
    const search = [
      'JId=123',
      'Jid=123',
      'JId=123&abc=456',
      'Jid=123&abc=456',
      'abc=123&JId=456',
      'abc=123&Jid=456',
      'abc=123&aJid=456',
      'abc=JId',
      '',
    ]

    test('with matched/unmatched referrers', () => {
      expect(
        _map(referrers, (referrer) => {
          global.document.referrer = referrer

          return utils.getEntryRoute()
        }),
      ).toMatchSnapshot()
    })

    test('with blank referrer and matched/unmatched URL param(s)', () => {
      global.document.referrer = ''

      expect(
        _map(search, (qs) => {
          global.history.replaceState({}, null, `/test.html?${qs}`)

          return utils.getEntryRoute()
        }),
      ).toMatchSnapshot()
    })
  })

  describe('formatFormikErrors', () => {
    const formikErrors = {
      loanAmount: 'Enter a loan amount.',
      mortgageLength: {
        months: 'Length of mortgage must include a month.',
        years: 'Length of mortgage must include a year.',
      },
      propertyValue: 'Enter a property value.',
      transactionType: 'Choose an option.',
      dependents: [{age: 'Enter an age'}, {age: 'Enter an age'}],
    }

    const analyticsErrors = [
      {cause: 'loanAmount', message: 'Enter a loan amount.'},
      {
        cause: 'mortgageLength.months',
        message: 'Length of mortgage must include a month.',
      },
      {
        cause: 'mortgageLength.years',
        message: 'Length of mortgage must include a year.',
      },
      {cause: 'propertyValue', message: 'Enter a property value.'},
      {cause: 'transactionType', message: 'Choose an option.'},
      {cause: 'dependents.0.age', message: 'Enter an age'},
      {cause: 'dependents.1.age', message: 'Enter an age'},
    ]

    it('should format formik errors object for the analytics message', () => {
      expect(utils.formatFormikErrors(formikErrors)).toMatchSnapshot()
      expect(utils.formatFormikErrors(formikErrors)).toEqual(analyticsErrors)
    })
  })
})
