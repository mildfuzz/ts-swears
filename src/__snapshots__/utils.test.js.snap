// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@analytics > utils formatFormikErrors should format formik errors object for the analytics message 1`] = `
Array [
  Object {
    "cause": "loanAmount",
    "message": "Enter a loan amount.",
  },
  Object {
    "cause": "mortgageLength.months",
    "message": "Length of mortgage must include a month.",
  },
  Object {
    "cause": "mortgageLength.years",
    "message": "Length of mortgage must include a year.",
  },
  Object {
    "cause": "propertyValue",
    "message": "Enter a property value.",
  },
  Object {
    "cause": "transactionType",
    "message": "Choose an option.",
  },
  Object {
    "cause": "dependents.0.age",
    "message": "Enter an age",
  },
  Object {
    "cause": "dependents.1.age",
    "message": "Enter an age",
  },
]
`;

exports[`@analytics > utils getDeviceData() returns device data with orientation & responsive state 1`] = `
Array [
  Array [
    "1x1",
    Object {
      "nbs_page_responsive_orientation": "square",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "300x300",
    Object {
      "nbs_page_responsive_orientation": "square",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "3x4",
    Object {
      "nbs_page_responsive_orientation": "portrait",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "768x1024",
    Object {
      "nbs_page_responsive_orientation": "portrait",
      "nbs_page_responsive_state": "medium",
    },
  ],
  Array [
    "9x16",
    Object {
      "nbs_page_responsive_orientation": "portrait",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "720x1280",
    Object {
      "nbs_page_responsive_orientation": "portrait",
      "nbs_page_responsive_state": "medium",
    },
  ],
  Array [
    "5x4",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "1280x1024",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "large",
    },
  ],
  Array [
    "8x5",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "1680x1050",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "large",
    },
  ],
  Array [
    "16x9",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "small",
    },
  ],
  Array [
    "1920x1080",
    Object {
      "nbs_page_responsive_orientation": "landscape",
      "nbs_page_responsive_state": "large",
    },
  ],
]
`;

exports[`@analytics > utils getEntryRoute() with blank referrer and matched/unmatched URL param(s) 1`] = `
Array [
  "banking app",
  "banking app",
  "banking app",
  "banking app",
  "banking app",
  "banking app",
  "external",
  "external",
  "external",
]
`;

exports[`@analytics > utils getEntryRoute() with matched/unmatched referrers 1`] = `
Array [
  "nationwide.co.uk",
  "nationwide.co.uk",
  "nationwide.co.uk",
  "nationwide.co.uk",
  "internet bank",
  "internet bank",
  "external",
  "external",
  "external",
]
`;

exports[`@analytics > utils getServiceData() returns service properties with mapped environment name 1`] = `
Array [
  Object {
    "nbs_environment": "local",
    "nbs_server": "http://localhost",
  },
  Object {
    "nbs_environment": "dev",
    "nbs_server": "http://localhost",
  },
  Object {
    "nbs_environment": "live",
    "nbs_server": "http://localhost",
  },
  Object {
    "nbs_environment": "unknown",
    "nbs_server": "http://localhost",
  },
]
`;

exports[`@analytics > utils load time measuring initial load start time 1`] = `
Object {
  "nbs-aws:application:overpayment:initial load": 101,
}
`;

exports[`@analytics > utils load time measuring loadEnd() sets & returns page load duration 1`] = `
Object {
  "withDate": Object {
    "end": Object {
      "bar": 10,
      "foo": 10,
      "nbs-aws:application:overpayment:initial load": 101,
    },
    "start": Object {
      "bar": 111,
      "foo": 10,
      "nbs-aws:application:overpayment:initial load": 101,
    },
  },
  "withPerf": Object {
    "end": Object {
      "foo": 10,
      "nbs-aws:application:overpayment:initial load": 101,
    },
    "start": Object {
      "foo": 222,
      "nbs-aws:application:overpayment:initial load": 101,
    },
  },
}
`;

exports[`@analytics > utils load time measuring loadStart() sets & returns page load start time 1`] = `
Object {
  "bar": 111,
  "foo": 222,
  "nbs-aws:application:overpayment:initial load": 101,
}
`;

exports[`@analytics > utils page categories getPageCategories() returns correctly constructed object 1`] = `
Array [
  Object {
    "nbs_sub_category_1": "nbs-aws:application",
    "nbs_sub_category_2": "nbs-aws:application:overpayment",
    "nbs_sub_category_3": "nbs-aws:application:overpayment",
  },
  Object {
    "nbs_sub_category_1": "nbs-aws:application",
    "nbs_sub_category_2": "nbs-aws:application:overpayment",
    "nbs_sub_category_3": "nbs-aws:application:overpayment:foo 1",
  },
  Object {
    "nbs_sub_category_1": "nbs-aws:application",
    "nbs_sub_category_2": "nbs-aws:application:overpayment",
    "nbs_sub_category_3": "nbs-aws:application:overpayment:bar 1",
  },
]
`;

exports[`@analytics > utils page name formatting getPageName() returns lowercase & trimmed page name string 1`] = `
Array [
  "nbs-aws:application:overpayment",
  "nbs-aws:application:overpayment",
  "nbs-aws:application:overpayment",
  "nbs-aws:application:overpayment",
  "nbs-aws:application:overpayment",
  "nbs-aws:application:overpayment:1:0:foo:bar",
]
`;

exports[`@analytics > utils page name formatting getPageNameSections() returns page name sections array with category prefixes 1`] = `
Array [
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    0,
  ],
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    null,
  ],
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    false,
  ],
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    undefined,
  ],
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    "",
  ],
  Array [
    "nbs-aws",
    "application",
    "OverPayment",
    0,
    1,
    "0",
    "foo",
    "bar",
  ],
]
`;
